File: examples/lua/client/protocols/MsgTestXX.lua
module("MsgTestXX", package.seeall)


function new()
	local tab_obj = {}
	setmetatable(tab_obj, {__index = MsgTestXX})
	tab_obj:init()
	return tab_obj
end

function init(self)
	self.id_u8 = 0
	self.id_f32 = {}
	self.id_op_u8 = 0
end


function encode(self)
	 local pack = packet.new()
	pack:writeU8(self.id_u8)
	id_f32_count = table.getn(self.id_f32)
	pack:writeU16(id_f32_count)
	for i = 1, id_f32_count do
		pack:writeF32(table.remove(self.id_f32))
	end
	if self.id_op_u8 ~= 0 then
	pack:writeU8(1)
		pack:writeU8(self.id_op_u8)
	else
		pack:writeU8(0)
	end
	return pack:readBytes()
end

function decode(self, pack)
	self.id_u8 = pack:readU8()
	id_f32_count = pack:readU16()
	for i = 1, id_f32_count do
		table.insert(self.id_f32, pack:readF32())
	end
	 local id_op_u8_flag = pack:readU8()
	if id_op_u8_flag then
		self.id_op_u8 = pack:readU8()
	end
	return self
end

function getBytes(self)
	return self:encode()
end

File: examples/golang/server/proto/msg_test_x_x.go
package proto

import (
	"packet"
)

type MsgTestXX struct {
	IdU8                     uint8
	IdF32                    []float32
	IdOpU8                   uint8
}

func (this *MsgTestXX) Encode() []byte {
	pack := packet.NewWriteBuff(64)

	pack.WriteUint8(this.IdU8)
	IdF32Count := uint16(len(this.IdF32))
	pack.WriteUint16(IdF32Count)
	for i := uint16(0); i < IdF32Count; i++ {
		pack.WriteFloat32(this.IdF32[i])
	}
	if this.IdOpU8 != 0 {
		pack.WriteUint8(1)
		pack.WriteUint8(this.IdOpU8)
	} else {
		pack.WriteUint8(0)
	}

	return pack.ReadBytes()
}

func MsgTestXXDecode(pack *packet.Packet) *MsgTestXX {
	msgTestXX := &MsgTestXX{}

	msgTestXX.IdU8 = pack.ReadUint8()
	IdF32Count := pack.ReadUint16()
	for ;IdF32Count > 0; IdF32Count-- {
		msgTestXX.IdF32 = append(msgTestXX.IdF32, pack.ReadFloat32())
	}
	IdOpU8Flag := pack.ReadUint8()
	if IdOpU8Flag == 1 {
		msgTestXX.IdOpU8 = pack.ReadUint8()
	}
	return msgTestXX
}
